model/DataQuery.php

```diff
@@ -640,6 +640,61 @@ public function leftJoin($table, $onClause, $alias = null, $order = 20, $paramet
 	}
 
 	/**
+	 * @param $relations string|array
+	 * @param $cb callable
+	 */
+	protected function relationWalk($relations, $cb)
+	{
+		if (is_string($relations)) $relations = explode('.', $relations);
+
+		$modelClass = $this->dataClass;
+
+		foreach ($relations as $relationName) {
+			/** @var DataObject $model */
+			$model = singleton($modelClass);
+			if ($relationClass = $model->hasOneComponent($relationName)) {
```

Comment:
Could you make use of [`getRelationClass()`](http://api.silverstripe.org/3.3/source-class-DataObject.html#_getRelationClass) here instead of doing these if statements?